compiler/src/iree/compiler/Codegen/Dialect/GPU/TransformExtensions/IREEGPUExtensionsOps.td

```diff
@@ -228,4 +228,38 @@ def FuseForallOp : Op<Transform_Dialect, "iree.fuse_forall",
   let cppNamespace = "mlir::iree_compiler::IREE::transform_dialect";
 }
 
+def FuseCollapseShapeWithForallOp : Op<Transform_Dialect, "iree.fuse_collapse_shape_with_forall",
+    [FunctionalStyleTransformOpTrait,
+     DeclareOpInterfaceMethods<MemoryEffectsOpInterface>,
+     DeclareOpInterfaceMethods<TransformOpInterface>,
+     ReportTrackingListenerFailuresOpTrait]> {
+  let description = [{
+    Fuses a consumer tensor.collapse_shape op into a producer scf.forall op.
+    The users of the block argument for the corresponding forall output operand
+    should be only a tensor.parallel_insert_slice op, and tensor.extract_slice
+    ops that extract an equivalent subset. After the fusion, the output of the
+    forall will be collapsed, and all users of this block arg will also be
+    collapsed. Additional tensor.expand_shape ops will be inserted after any
+    tensor.extract_slice users inside the forall so that types match. Similarly,
```

Comment:
hmmm I would expect the extract_slice fusion case to work like this
```mlir
%dest = ...
%loop = scf.forall shared_outs(%init = %dest) {
  %slice = tensor.extract_slice %init
  %prod = ...
  tensor.parallel_insert_slice %prod into %init
}
%collapse = tensor.extract_slice %loop
```
```mlir
%dest = ...
%empty = tensor.empty(<extract_slice_sizes>)
%loop = scf.forall shared_outs(%init = %empty) {
  %slice = tensor.extract_slice %dest
  %prod = ...
  %fused_slice = tensor.extract_slice
  tensor.parallel_insert_slice %fused_slice into %init
}
```
Because that first `%slice` on the loop iter_arg can't be modified by the extract_slice fusion.

Reply:
This looks wrong to me. We can only do this if `%dest` is a tensor.empty. Otherwise, we are ignoring the initialized values of the shared_outs. Also, in the case you described above, the `%slice` still needs to be modified because `%init` is the size of the sliced tensor. The way my pattern works is like this:
```mlir
%dest = ...
%loop = scf.forall shared_outs(%init = %dest) {
  %slice = tensor.extract_slice %init
  %prod = ...
  tensor.parallel_insert_slice %prod into %init
}
%collapse = tensor.extract_slice %loop
```
```mlir
%dest = ...
%new_dest = tensor.extract_slice %dest ...
%loop = scf.forall shared_outs(%init = %new_dest) {
  %slice = tensor.extract_slice %dest [<clamped sizes>]
  %pad = tensor.pad %slice ...
  %prod = ...
  %fused_slice = tensor.extract_slice
  tensor.parallel_insert_slice %fused_slice into %init
}
```

Reply:
Good point, then the final IR should look like this
```mlir
%dest = ...
%new_dest = tensor.extract_slice %dest ...
%loop = scf.forall shared_outs(%init = %new_dest) {
  %slice = tensor.extract_slice %dest
  %prod = ...
  %fused_slice = tensor.extract_slice
  tensor.parallel_insert_slice %fused_slice into %init
}
```
so instead of an empty we have an extract_slice on the original destination. Inserting a pad is invalid because there is no reasonable way to determine the padding value.

Reply:
It might help to take a look at how TilingInterface based consumer fusion is done upstream. The `tensor.extract_slice` on the new loop inits comes from the fact that it requires producers are DPS: https://github.com/llvm/llvm-project/blob/9b64811e27f8377b77f8561cf678c9d2f2d67393/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp#L2033

Because `tensor.extract_slice` is not DPS, we won't end up with any `tensor.extract_slice` on the loop iter args. Updating an existing destination doesn't work because consumers of existing destinations are free to read and use whatever values they want. Also FWIW no other consumer fusions handle the case where an `scf.forall` does not overwrite its whole destination (in fact if we are generating `scf.forall` ops like that it's likely a major correctness footgun).

Reply:
I think this will create a new buffer for the whole destination though (?)

Reply:
> Good point, then the final IR should look like this
> 
> ```mlir
> %dest = ...
> %new_dest = tensor.extract_slice %dest ...
> %loop = scf.forall shared_outs(%init = %new_dest) {
>   %slice = tensor.extract_slice %dest
>   %prod = ...
>   %fused_slice = tensor.extract_slice
>   tensor.parallel_insert_slice %fused_slice into %init
> }
> ```
> 
> so instead of an empty we have an extract_slice on the original destination. Inserting a pad is invalid because there is no reasonable way to determine the padding value.

Following up after an offline discussion. This is the IR we want to produce, and we expect that there should be a `linalg.copy` (or other similar op) following the tensor.extract_slice that is being fused. The copy will then be fused into the loop, and update the destination in a way that bufferization is happy with. Any input IR to the pattern that does not have a DPS op following the tensor.extract_slice is problematic to begin with, since it does not obey DPS, so it should not be the job of this pattern to create happy IR for bufferization. Instead, it is expected that the consumer of the extract_slice can fix up the inits after being fused.

I will refactor this PR and the next PR to get rid of all that extract_slice matching, but before this lands I need to fix some patterns upstream that cause linalg.copy ops to get folded (which loses DPS).